File: backend/app/models/aiops.py
"""
AIOpsAgent, AgentTask, AgentRecommendation, and AgentExecution models for Autonomous AIOps Agent.
"""

import uuid
from datetime import datetime
from sqlalchemy import String, Text, ForeignKey, JSON, Float, Integer, Boolean, DateTime
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base_class import Base
from app.models.mixins import UUIDMixin, TimestampMixin


class AIOpsAgent(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "aiops_agents"

    agent_name: Mapped[str] = mapped_column(String(255), nullable=False, default="CloudPulse Autonomous Core")
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="Autonomous")  # Active, Autonomous, Paused
    current_phase: Mapped[str] = mapped_column(String(50), nullable=False, default="Observe")  # Observe, Detect, Analyze, Plan, Recommend, Verify
    health_status: Mapped[str] = mapped_column(String(50), nullable=False, default="Healthy")  # Healthy, Degraded, Anomalous
    last_observation_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    tasks: Mapped[list["AgentTask"]] = relationship("AgentTask", back_populates="agent", cascade="all, delete-orphan")
    recommendations: Mapped[list["AgentRecommendation"]] = relationship("AgentRecommendation", back_populates="agent", cascade="all, delete-orphan")


class AgentTask(UUIDMixin, Base):
    __tablename__ = "agent_tasks"

    agent_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("aiops_agents.id", ondelete="CASCADE"), nullable=False)
    task_name: Mapped[str] = mapped_column(String(255), nullable=False)
    target_system: Mapped[str] = mapped_column(String(100), nullable=False)  # Metrics, Logs, Traces, Security, Cost
    status: Mapped[str] = mapped_column(String(50), default="Completed")  # Pending, In_Progress, Completed, Failed
    started_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)
    completed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=True)

    agent: Mapped["AIOpsAgent"] = relationship("AIOpsAgent", back_populates="tasks")


class AgentRecommendation(UUIDMixin, TimestampMixin, Base):
    __tablename__ = "agent_recommendations"

    agent_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("aiops_agents.id", ondelete="CASCADE"), nullable=False)
    title: Mapped[str] = mapped_column(String(500), nullable=False)
    category: Mapped[str] = mapped_column(String(100), nullable=False, default="Anomaly_Detection")  # Anomaly_Detection, Root_Cause, Performance, Capacity_Planning, Risk_Prediction, Cost_Optimization, Security_Correlation
    priority: Mapped[str] = mapped_column(String(50), nullable=False, default="P1")  # P0, P1, P2, P3
    executive_summary: Mapped[str] = mapped_column(Text, nullable=False)
    root_cause: Mapped[str] = mapped_column(Text, nullable=True)
    business_impact: Mapped[str] = mapped_column(Text, nullable=True)
    recommended_actions: Mapped[list] = mapped_column(JSON, nullable=True, default=list)
    automation_candidates: Mapped[list] = mapped_column(JSON, nullable=True, default=list)
    confidence_score: Mapped[float] = mapped_column(Float, default=0.95)
    expected_recovery_time: Mapped[str] = mapped_column(String(100), default="10 mins")
    status: Mapped[str] = mapped_column(String(50), nullable=False, default="Pending_Approval")  # Pending_Approval, Approved, Rejected, Executed

    agent: Mapped["AIOpsAgent"] = relationship("AIOpsAgent", back_populates="recommendations")
    executions: Mapped[list["AgentExecution"]] = relationship("AgentExecution", back_populates="recommendation", cascade="all, delete-orphan")


class AgentExecution(UUIDMixin, Base):
    __tablename__ = "agent_executions"

    recommendation_id: Mapped[uuid.UUID] = mapped_column(UUID(as_uuid=True), ForeignKey("agent_recommendations.id", ondelete="CASCADE"), nullable=False)
    action_taken: Mapped[str] = mapped_column(Text, nullable=False)
    approved_by: Mapped[str] = mapped_column(String(255), nullable=False, default="AIOps Autonomous Controller")
    status: Mapped[str] = mapped_column(String(50), default="Completed")
    execution_logs: Mapped[list] = mapped_column(JSON, nullable=True, default=list)
    executed_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), nullable=False)

    recommendation: Mapped["AgentRecommendation"] = relationship("AgentRecommendation", back_populates="executions")

File: frontend/src/types/aiops.ts
/**
 * TypeScript Type Definitions for Autonomous AIOps Agent & AI Operations Center
 */

export interface AgentTask {
  id: string;
  agent_id: string;
  task_name: string;
  target_system: string;
  status: "Pending" | "In_Progress" | "Completed" | "Failed";
  started_at: string;
  completed_at?: string;
}

export interface AgentExecution {
  id: string;
  recommendation_id: string;
  action_taken: string;
  approved_by: string;
  status: string;
  execution_logs: string[];
  executed_at: string;
}

export interface AgentRecommendation {
  id: string;
  agent_id: string;
  title: string;
  category:
    | "Anomaly_Detection"
    | "Root_Cause"
    | "Performance"
    | "Capacity_Planning"
    | "Risk_Prediction"
    | "Cost_Optimization"
    | "Security_Correlation";
  priority: "P0" | "P1" | "P2" | "P3";
  executive_summary: string;
  root_cause?: string;
  business_impact?: string;
  recommended_actions: string[];
  automation_candidates: string[];
  confidence_score: number;
  expected_recovery_time: string;
  status: "Pending_Approval" | "Approved" | "Rejected" | "Executed";
  created_at: string;
  executions: AgentExecution[];
}

export interface AIOpsAgentStatus {
  id: string;
  agent_name: string;
  status: "Active" | "Autonomous" | "Paused";
  current_phase: "Observe" | "Detect" | "Analyze" | "Plan" | "Recommend" | "Verify";
  health_status: "Healthy" | "Degraded" | "Anomalous";
  last_observation_at: string;
  total_recommendations: number;
  pending_approvals: number;
  active_automations: number;
  tasks: AgentTask[];
}

export interface AIOpsListResponse {
  items: AgentRecommendation[];
  total: number;
  page: number;
  size: number;
  pages: number;
}
